src/hotspot/share/gc/g1/g1CardSetMemory.hpp

```diff
@@ -84,44 +84,30 @@ typedef G1SegmentedArrayFreeList<mtGCCardSet> G1CardSetFreeList;
 // Since it is expected that every CardSet (and in extension each region) has its
 // own set of allocators, there is intentionally no padding between them to save
 // memory.
-template <class Slot>
 class G1CardSetAllocator {
-  // G1CardSetSegment management.
-
-  typedef G1SegmentedArray<Slot, mtGCCardSet> SegmentedArray;
   // G1CardSetContainer slot management within the G1CardSetSegments allocated
   // by this allocator.
-
-  SegmentedArray _segmented_array;
+  G1SegmentedArray<mtGCCardSet> _segmented_array;
   FreeListAllocator _free_slots_list;
 
 public:
   G1CardSetAllocator(const char* name,
-                     const G1CardSetAllocOptions* alloc_options,
-                     G1CardSetFreeList* free_segment_list);
+                         const G1CardSetAllocOptions* alloc_options,
+                         G1CardSetFreeList* free_segment_list);
   ~G1CardSetAllocator();
 
-  Slot* allocate();
-  void free(Slot* slot);
+  void* allocate();
+  void free(void* slot);
 
   // Deallocate all segments to the free segment list and reset this allocator. Must
   // be called in a globally synchronized area.
   void drop_all();
 
-  size_t mem_size() const {
-    return sizeof(*this) +
-      _segmented_array.num_segments() * sizeof(G1CardSetSegment) +
-      _segmented_array.num_available_slots() * _segmented_array.slot_size();
-  }
+  inline size_t mem_size() const;
 
-  size_t wasted_mem_size() const {
-    uint num_wasted_slots = _segmented_array.num_available_slots() -
-                            _segmented_array.num_allocated_slots() -
-                            (uint)_free_slots_list.pending_count();
-    return num_wasted_slots * _segmented_array.slot_size();
-  }
+  inline size_t wasted_mem_size() const;
 
-  inline uint num_segments() { return _segmented_array.num_segments(); }
+  inline uint num_segments() const;
```

Comment:
Do `mem_size` and `wasted_size` need to be inline?  They don't seem like they should be performance critical.  I'm not sure about `num_segments` because I haven't checked uses.